File: rudder-android-client/rudder-sdk-core/src/main/java/com/rudderlabs/android/sdk/core/RudderElementBuilder.java
package com.rudderlabs.android.sdk.core;

import java.util.Map;

public class RudderElementBuilder {
    private String eventName = null;

    public RudderElementBuilder setEventName(String eventName) {
        this.eventName = eventName;
        return this;
    }

    private String userId = null;

    public RudderElementBuilder setUserId(String userId) {
        this.userId = userId;
        return this;
    }

    private RudderProperty property;

    public RudderElementBuilder setProperty(RudderProperty property) {
        this.property = property;
        return this;
    }

    public RudderElementBuilder setProperty(RudderPropertyBuilder builder) throws RudderException {
        this.property = builder.build();
        return this;
    }

    public RudderElementBuilder setProperty(Map<String, Object> map) {
        property = new RudderProperty();
        property.setProperty(map);
        return this;
    }

    public RudderElement build() {
        RudderElement event = new RudderElement();
        event.setUserId(userId);
        event.setEventName(eventName);
        event.setProperty(property);
        return event;
    }
}

File: old-references/rudder-ios-client/RudderClient/Event/RudderElementBuilder.swift
//
//  RudderEventBuilder.swift
//  RudderSample
//
//  Created by Arnab Pal on 04/07/19.
//  Copyright © 2019 Arnab Pal. All rights reserved.
//

import Foundation

class RudderElementBuilder {
    internal var property: RudderProperty = RudderProperty()
    func setPropertyBuilder(builder: RudderPropertyBuilder) throws -> RudderElementBuilder {
        self.property = try builder.build()
        return self
    }
    
    func setProperty(property: RudderProperty) -> RudderElementBuilder {
        self.property = property
        return self
    }
    
    internal var event: String = ""
    func setEvent(event: String) -> RudderElementBuilder {
        self.event = event
        return self
    }
    
    internal var channel: String = ""
    func setChannel(channel: String) -> RudderElementBuilder {
        self.channel = channel
        return self
    }
    
    internal var userId: String = ""
    func setUserId(userId: String) -> RudderElementBuilder {
        self.userId = userId
        return self
    }
    
    func build() -> RudderElement {
        return RudderElement().fromBuilder(builder: self)
    }
}
